src/rust/pipeline-ingress/src/main.rs

```diff
@@ -0,0 +1,175 @@
+        let tenant_id = request.tenant_id;
+        let event_source_id = request.event_source_id;
+        // TODO: trace_id should be generated at the edge. This service is
+        // currently "the edge" but that won't be true forever. When there is an
+        // actual edge service, that service should be responsible for
+        // generating the trace_id.
+        let trace_id = Uuid::new_v4();
+        self.producer
+            .send(Envelope::new(
+                Metadata::new(
+                    tenant_id,
+                    trace_id,
+                    0,
+                    created_time,
+                    last_updated_time,
+                    event_source_id,
+                ),
+                RawLog::new(request.log_event),
+            ))
+            .await?;
+
+        Ok(PublishRawLogResponse::ok())
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Error)]
+enum ConfigurationError {
+    #[error("failed to configure kafka client {0}")]
+    Kafka(#[from] KafkaConfigurationError),
+
+    #[error("failed to configure gRPC server {0}")]
+    Server(#[from] ServerConfigurationError),
+
+    #[error("missing environment variable {0}")]
+    EnvironmentVariable(#[from] VarError),
+
+    #[error("failed to bind socket address {0}")]
+    SocketError(#[from] std::io::Error),
+
+    #[error("failed to parse integer value {0}")]
+    ParseInt(#[from] ParseIntError),
+}
+
+#[tracing::instrument(err)]
+async fn handler() -> Result<(), ConfigurationError> {
+    let socket_address = std::env::var("PIPELINE_INGRESS_BIND_ADDRESS")?;
+    let healthcheck_polling_interval_ms =
+        std::env::var("PIPELINE_INGRESS_HEALTHCHECK_POLLING_INTERVAL_MS")?.parse()?;
+
+    tracing::info!("configuring kafka producer");
+    let producer: Producer<Envelope<RawLog>> = Producer::new("raw-logs")?;
+    tracing::info!("kafka producer configured successfully");
+
+    tracing::info!("configuring gRPC server");
+    let (server, _shutdown_tx) = PipelineIngressServer::new(
+        IngressApi::new(producer),
+        TcpListener::bind(socket_address).await?,
+        || async { Ok(HealthcheckStatus::Serving) }, // FIXME: this is garbage
+        Duration::from_millis(healthcheck_polling_interval_ms),
+    );
+    tracing::info!("gRPC server configured successfully");
+
+    tracing::info!("starting gRPC server");
+    Ok(server.serve().await?)
+}
+
+#[tokio::main]
+async fn main() -> Result<(), ConfigurationError> {
+    let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::stdout());
+
+    // initialize json logging layer
+    let log_layer = tracing_subscriber::fmt::layer()
+        .json()
+        .with_writer(non_blocking);
+
+    // initialize tracing layer
+    global::set_text_map_propagator(TraceContextPropagator::new());
+    let tracer = opentelemetry_jaeger::new_pipeline()
+        .with_service_name("pipeline-ingress")
+        .install_batch(opentelemetry::runtime::Tokio)
+        .unwrap();
```

Comment:
Thanks, I missed this when I copy-pasta'd from the `grapl-config` crate, which I deliberately do not depend on here because it would drag in a whole bunch of heavy dependencies I don't want or need.

Reply:
fixed